include/boost/url/grammar/parse.hpp

```diff
@@ -77,216 +63,166 @@ parse(
-         ++it;
-       }
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (grammar::parse(it, end, ec, detail::alpha_string_rule{alpha_l}, '.',
-                    detail::alpha_string_rule{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
-   }
-   @endcode
+    If an error occurs for any of the rules in the
+    parameter pack, the entire parse operation fails
+    but the chars from the substring are still consumed.
 
-   @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
+    As other overloads of this function, `using grammar::parse`
+    is required to call this function appropriately with for
+    argument-dependent lookup for other grammar rules, and
+    calls to `parse` should be unqualified.
 
-   @see https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form
+    @par Example
+    @code
+    using grammar::parse;
+    if (parse(it, end, ec, r1, r2, r3)) {
+      std::cout << "Range [it, end) parsed successfully\n";
+    }
+    @endcode
 
-        @ref boost::urls::error
+    @par Exception Safety
 
-   @return `true` if the range initial chars match the elements successfully,
-   `false` otherwise
+      Defined by the types of the rule objects.
 
-   @param[in,out] it An iterator to the first element in the range.
-   At the end of this function, @ref it points to one past the last
-   element parsed in the range.
+    @return `true` if the range initial chars match the elements successfully
+
+    @param it An iterator to the first element in the range.
+    At the end of this function, `it` points to one past the last
+    element parsed in the range.
+
+    @param end An iterator to the end of the range
 
-   @param[in] end An iterator to one past the last element in the range
+    @param ec Set to the error, if any occurred
 
-   @param[out] ec The error code this function should set if it fails
+    @param r1 First grammar rule object
 
-   @param[out] args BNF elements to be parsed
+    @param r2 Second grammar rule object
+
+    @param rs Extra grammar rule objects
 
 */
-#ifdef BOOST_URL_DOCS
-template<class... Args>
-bool parse( char const*& it, char const* end, Args&&... args);
-#else
 template<
-    class T0,
-    class T1,
-    class... Tn>
+    class R0,
+    class R1,
+    class... Rs
+>
 bool
 parse(
     char const*& it,
     char const* end,
     error_code& ec,
-    T0&& t0,
-    T1&& t1,
-    Tn&&... tn);
-#endif
+    R0&& r1,
+    R1&& r2,
+    Rs&&... rs
```

Comment:
I use `Rn&& rn` but `Rs&& rs` works too I guess

Reply:
I thought about `Rn` too, the problem is `Rn` usually indicates the last element instead of a sequence of elements. Of course, that makes little difference to anyone who understands variadic templates but still...